File: Backend/agent audit flow -copy/agents/retrieval_agent.py
import json
from pathlib import Path

DB_PATH = Path(__file__).resolve().parent.parent / "data" / "official_db.json"


def load_official_db(db_path=DB_PATH):
    with open(db_path, "r", encoding="utf-8") as f:
        return json.load(f)


def retrieval_agent(query: dict, db_path=DB_PATH):
    """
    Input query example:
    { "project_id": "MH-RD-001" }
    OR
    { "area": "Andheri East" }

    Output:
    official record JSON or None
    """
    db = load_official_db(db_path)

    project_id = query.get("project_id")
    area = query.get("area")

    for record in db:
        if project_id and record.get("project_id") == project_id:
            return record

        if area and record.get("location", {}).get("area", "").lower() == area.lower():
            return record

    return None

File: Backend/src/features/agentic-audit/agents/retrieval_agent.js
const fs = require('fs');
const path = require('path');

const officialDbPath = path.join(__dirname, '../data/official_db.json');

const delay = (ms) => new Promise(resolve => setTimeout(resolve, ms));

const retrievalAgent = async (projectId, location) => {
  const startTime = Date.now();
  
  // Simulate 1-2 second delay
  await delay(1000 + Math.random() * 1000);
  
  // Load official database
  const officialDb = JSON.parse(fs.readFileSync(officialDbPath, 'utf8'));
  
  let project = null;
  
  if (projectId) {
    project = officialDb.find(p => p.project_id === projectId);
  } else if (location) {
    // Find project by location proximity (within 0.01 degrees)
    project = officialDb.find(p => {
      const latDiff = Math.abs(p.location.lat - location.lat);
      const lngDiff = Math.abs(p.location.lng - location.lng);
      return latDiff < 0.01 && lngDiff < 0.01;
    });
  }
  
  const endTime = Date.now();
  const duration = ((endTime - startTime) / 1000).toFixed(2);
  
  return {
    success: !!project,
    project: project,
    log: {
      agent: 'retrieval_agent',
      action: 'fetch_official_records',
      duration: `${duration}s`,
      timestamp: new Date().toISOString(),
      message: project 
        ? `Found official record for project ${project.project_id}: ${project.name}`
        : 'No matching official record found'
    }
  };
};

module.exports = retrievalAgent;
